pkg/util/hostname.go

```diff
@@ -159,6 +159,25 @@ func GetHostnameData(ctx context.Context) (HostnameData, error) {
 	var err error
 	var provider string
 
+	if config.Datadog.GetBool("hostname_force_ec2_instance_id") {
+		if getEC2Hostname := hostname.GetProvider("ec2"); getEC2Hostname != nil {
+			hostName, err = getValidEC2Hostname(ctx, getEC2Hostname)
+		} else {
+			err = fmt.Errorf("ec2 hostname provider is not enabled")
+		}
+
+		if err == nil {
+			return saveHostnameData(cacheHostnameKey, hostName, "aws"), nil
+		}
+
+		expErr := new(expvar.String)
+		expErr.Set(err.Error())
+		hostnameErrors.Set("forced EC2 hostname", expErr)
+		// We want to prevent the agent from starting with another hostname if 'hostname_force_ec2_instance_id'
+		// is enabled.
+		return HostnameData{}, fmt.Errorf("'hostname_force_ec2_instance_id' is enable and EC2 instanceID could not be fetched: %s", err)
+	}
```

Comment:
I think this block should be combined with the [EC2 hostname eval](https://github.com/DataDog/datadog-agent/blob/main/pkg/util/hostname.go#L272-L288) since both blocks are pretty much doing the same thing.

Reply:
We _really_ need a matrix test for this :(

Reply:
> that's the issue of this whole function: lot of tricky edge case. We actually want to force EC2 so we want to check is first thing before we use other providers like the configuration.

I think the fall-through for the new logic should still be the same just that there will be another case tacked on to `if ecs.IsECSInstance() || ec2.IsDefaultHostname(hostName)` case. I believe (if I'm reading the flow correctly) that even if there's a FQDN and OS hostname, we don't early return until we go through the EC2 logic which will should override the hostname as expected.

I guess my overall point is that we use this fall-through for the same logic to use instance IDs instead of hostnames already (when it's a default EC2 hostname) so I wouldn't expect that we would be introducing any gotcha/unexpected new behaviors with the logic change occurring in the EC2 section.

Reply:
This would mean that we don't enforce EC2 instance ID but only prioritize it above OS/FQDN. I think the config name is misleading in that case. WDYT ?

Reply:
After our discussion I merged it with the current EC2 logic and updated the configuration name